e2e: Test Upgrades of Rancher while Elemental is installed

It would be useful to validate that the Elemental operator and downstream clusters survive an upgrade of Rancher Manager. 

The basic flow that I'm thinking is:
- Install and configure Rancher & Elemental
- Register nodes
- Create cluster
- Upgrade Rancher 
- Confirm that clusters are still attached and in healthy state
- Confirm that Elemental operator is not producing any unexpected logging
- Register new node to confirm that registration is still healthy

I tried the upgrade process without Elemental stack and it worked, then tried again with Elemental and hit the same issue:
``` 
2023/03/31 08:10:59 [ERROR] [CAPI] Unable to retrieve Node status: failed to create cluster accessor: failed to get lock for cluster: cluster is locked already
2023/03/31 08:11:01 [ERROR] [CAPI] Unable to retrieve Node status: failed to create cluster accessor: failed to get lock for cluster: cluster is locked already
2023/03/31 08:11:03 [ERROR] error parsing azure-group-cache-size, skipping update strconv.Atoi: parsing "": invalid syntax
2023/03/31 08:11:03 [ERROR] error syncing 'fleet-default/mycluster-managed-system-upgrade-controller': handler mcc-bundle: no chart version found for system-upgrade-controller-102.0.0+up0.4.0, requeuing
2023/03/31 08:11:04 [ERROR] error syncing 'fleet-default/mycluster-managed-system-upgrade-controller': handler mcc-bundle: no chart version found for system-upgrade-controller-102.0.0+up0.4.0, requeuing
2023/03/31 08:11:04 [ERROR] error syncing 'fleet-default/mycluster-managed-system-upgrade-controller': handler mcc-bundle: no chart version found for system-upgrade-controller-102.0.0+up0.4.0, requeuing
2023/03/31 08:11:04 [ERROR] error syncing 'fleet-default/mycluster-managed-system-upgrade-controller': handler mcc-bundle: no chart version found for system-upgrade-controller-102.0.0+up0.4.0, requeuing
2023/03/31 08:11:04 [ERROR] error syncing 'fleet-default/mycluster-managed-system-upgrade-controller': handler mcc-bundle: no chart version found for system-upgrade-controller-102.0.0+up0.4.0, requeuing
2023/03/31 08:11:04 [ERROR] available chart version (101.0.1+up0.5.1) for fleet is less than the min version (102.0.0+up0.6.0-rc.5) 
2023/03/31 08:11:04 [ERROR] Failed to find system chart fleet will try again in 5 seconds: no chart name found
2023/03/31 08:11:05 [ERROR] error syncing 'fleet-default/mycluster-managed-system-upgrade-controller': handler mcc-bundle: no chart version found for system-upgrade-controller-102.0.0+up0.4.0, requeuing
2023/03/31 08:11:05 [ERROR] error syncing 'fleet-default/mycluster-managed-system-upgrade-controller': handler mcc-bundle: no chart version found for system-upgrade-controller-102.0.0+up0.4.0, requeuing
2023/03/31 08:11:05 [ERROR] error syncing 'fleet-default/mycluster-managed-system-upgrade-controller': handler mcc-bundle: no chart version found for system-upgrade-controller-102.0.0+up0.4.0, requeuing
2023/03/31 08:11:06 [WARNING] [clean-catalog-orphan-bindings] Error when deleting role global-catalog, roles.rbac.authorization.k8s.io "global-catalog" not found
2023/03/31 08:11:06 [WARNING] [CleanupOrphanCatalogBindingsDone] error during orphan binding cleanup: roles.rbac.authorization.k8s.io "global-catalog" not found
2023/03/31 08:11:06 [ERROR] failed to cleanup orphan catalog bindings
2023/03/31 08:11:06 [ERROR] error syncing 'fleet-default/mycluster-managed-system-upgrade-controller': handler mcc-bundle: no chart version found for system-upgrade-controller-102.0.0+up0.4.0, requeuing
2023/03/31 08:11:08 [ERROR] [CAPI] Reconciler error: failed to add &Node{ObjectMeta:{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []},Spec:NodeSpec{PodCIDR:,DoNotUseExternalID:,ProviderID:,Unschedulable:false,Taints:[]Taint{},ConfigSource:nil,PodCIDRs:[],},Status:NodeStatus{Capacity:ResourceList{},Allocatable:ResourceList{},Phase:,Conditions:[]NodeCondition{},Addresses:[]NodeAddress{},DaemonEndpoints:NodeDaemonEndpoints{KubeletEndpoint:DaemonEndpoint{Port:0,},},NodeInfo:NodeSystemInfo{MachineID:,SystemUUID:,BootID:,KernelVersion:,OSImage:,ContainerRuntimeVersion:,KubeletVersion:,KubeProxyVersion:,OperatingSystem:,Architecture:,},Images:[]ContainerImage{},VolumesInUse:[],VolumesAttached:[]AttachedVolume{},Config:nil,},} watch on cluster fleet-default/mycluster: failed to create cluster accessor: error creating dynamic rest mapper for remote cluster "fleet-default/mycluster": Get "https://10.43.71.189/k8s/clusters/c-m-fd8ggfds/api?timeout=10s": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
2023/03/31 08:11:08 [ERROR] failed to call leader func: dynamicschemas.management.cattle.io "machineinventoryselectortemplate" not found
2023/03/31 08:11:09 [ERROR] error syncing 'fleet-default/mycluster-managed-system-upgrade-controller': handler mcc-bundle: no chart version found for system-upgrade-controller-102.0.0+up0.4.0, requeuing
2023/03/31 08:11:10 [ERROR] error syncing 'fleet-default/mycluster': handler provisioning-cluster-change: dynamicschemas.management.cattle.io "machineinventoryselectortemplate" not found, requeuing
2023/03/31 08:11:10 [ERROR] error syncing 'fleet-default/mycluster': handler provisioning-cluster-change: dynamicschemas.management.cattle.io "machineinventoryselectortemplate" not found, requeuing
2023/03/31 08:11:10 [ERROR] available chart version (101.0.1+up0.5.1) for fleet is less than the min version (102.0.0+up0.6.0-rc.5) 
2023/03/31 08:11:10 [ERROR] Failed to find system chart fleet will try again in 5 seconds: no chart name found
```

> I tried the upgrade process without Elemental stack and it worked, then tried again with Elemental and hit the same issue:
> 
> ```
> [...]
> 2023/03/31 08:11:08 [ERROR] failed to call leader func: dynamicschemas.management.cattle.io "machineinventoryselectortemplate" not found
> [...]
> ```

This is the issue tracked in #780

--- a/.github/workflows/cli-k3s-os-upgrade-rancher_latest.yaml
+++ b/.github/workflows/cli-k3s-os-upgrade-rancher_latest.yaml
@@ -20,8 +20,7 @@ jobs:
       cluster_name: cluster-k3s
       iso_to_test: https://download.opensuse.org/repositories/isv:/Rancher:/Elemental:/Stable:/Teal53/media/iso/elemental-teal.x86_64.iso
       k8s_version_to_provision: v1.25.7+k3s1
-      # Force to only deploy the first 3 nodes
-      node_number: 3
+      node_number: 5
       operator_upgrade: oci://registry.opensuse.org/isv/rancher/elemental/dev/charts/rancher
       operator_repo: oci://registry.opensuse.org/isv/rancher/elemental/stable/charts/rancher
       rancher_version: latest/devel

--- a/.github/workflows/cli-k3s-os-upgrade-rancher_stable.yaml
+++ b/.github/workflows/cli-k3s-os-upgrade-rancher_stable.yaml
@@ -20,10 +20,10 @@ jobs:
       cluster_name: cluster-k3s
       iso_to_test: https://download.opensuse.org/repositories/isv:/Rancher:/Elemental:/Stable:/Teal53/media/iso/elemental-teal.x86_64.iso
       k8s_version_to_provision: v1.25.7+k3s1
-      # Force to only deploy the first 3 nodes
-      node_number: 3
+      node_number: 5
       operator_upgrade: oci://registry.opensuse.org/isv/rancher/elemental/dev/charts/rancher
       operator_repo: oci://registry.opensuse.org/isv/rancher/elemental/stable/charts/rancher
+      rancher_upgrade: latest/devel
       rancher_version: stable/latest
       upgrade_channel_list: registry.opensuse.org/isv/rancher/elemental/dev/teal53/15.4/rancher/elemental-teal-dev-channel/5.3:latest
       upgrade_image: registry.opensuse.org/isv/rancher/elemental/dev/teal53/15.4/rancher/elemental-teal/5.3:latest

--- a/.github/workflows/cli-k3s-os-upgrade.yaml
+++ b/.github/workflows/cli-k3s-os-upgrade.yaml
@@ -12,6 +12,10 @@ on:
         description: ISO to test
         default: https://download.opensuse.org/repositories/isv:/Rancher:/Elemental:/Stable:/Teal53/media/iso/elemental-teal.x86_64.iso
         type: string
+      rancher_upgrade:
+        description: Rancher Manager channel/version to upgrade to
+        default: latest/devel
+        type: string
       rancher_version:
         description: Rancher Manager channel/version to use for installation
         default: stable/latest
@@ -38,10 +42,10 @@ jobs:
       destroy_runner: ${{ inputs.destroy_runner }}
       iso_to_test: ${{ inputs.iso_to_test }}
       k8s_version_to_provision: v1.25.7+k3s1
-      # Force to only deploy the first 3 nodes
-      node_number: 3
+      node_number: 5
       operator_upgrade: oci://registry.opensuse.org/isv/rancher/elemental/${{ inputs.upgrade_os_channel }}/charts/rancher
       operator_repo: oci://registry.opensuse.org/isv/rancher/elemental/stable/charts/rancher
+      rancher_upgrade: ${{ inputs.rancher_upgrade }}
       rancher_version: ${{ inputs.rancher_version }}
       upgrade_channel_list: registry.opensuse.org/isv/rancher/elemental/dev/teal53/15.4/rancher/elemental-teal-dev-channel/5.3:latest
       upgrade_image: registry.opensuse.org/isv/rancher/elemental/${{ inputs.upgrade_os_channel }}/teal53/15.4/rancher/elemental-teal/5.3:latest

--- a/.github/workflows/cli-rke2-os-upgrade-rancher_latest.yaml
+++ b/.github/workflows/cli-rke2-os-upgrade-rancher_latest.yaml
@@ -21,8 +21,7 @@ jobs:
       cluster_name: cluster-rke2
       iso_to_test: https://download.opensuse.org/repositories/isv:/Rancher:/Elemental:/Stable:/Teal53/media/iso/elemental-teal.x86_64.iso
       k8s_version_to_provision: v1.25.7+rke2r1
-      # Force to only deploy the first 3 nodes
-      node_number: 3
+      node_number: 5
       operator_upgrade: oci://registry.opensuse.org/isv/rancher/elemental/dev/charts/rancher
       operator_repo: oci://registry.opensuse.org/isv/rancher/elemental/stable/charts/rancher
       rancher_version: latest/devel

--- a/.github/workflows/cli-rke2-os-upgrade-rancher_stable.yaml
+++ b/.github/workflows/cli-rke2-os-upgrade-rancher_stable.yaml
@@ -21,10 +21,10 @@ jobs:
       cluster_name: cluster-rke2
       iso_to_test: https://download.opensuse.org/repositories/isv:/Rancher:/Elemental:/Stable:/Teal53/media/iso/elemental-teal.x86_64.iso
       k8s_version_to_provision: v1.25.7+rke2r1
-      # Force to only deploy the first 3 nodes
-      node_number: 3
+      node_number: 5
       operator_upgrade: oci://registry.opensuse.org/isv/rancher/elemental/dev/charts/rancher
       operator_repo: oci://registry.opensuse.org/isv/rancher/elemental/stable/charts/rancher
+      rancher_upgrade: latest/devel
       rancher_version: stable/latest
       upgrade_channel_list: registry.opensuse.org/isv/rancher/elemental/dev/teal53/15.4/rancher/elemental-teal-dev-channel/5.3:latest
       upgrade_image: registry.opensuse.org/isv/rancher/elemental/dev/teal53/15.4/rancher/elemental-teal/5.3:latest

--- a/.github/workflows/cli-rke2-os-upgrade.yaml
+++ b/.github/workflows/cli-rke2-os-upgrade.yaml
@@ -12,6 +12,10 @@ on:
         description: ISO to test
         default: https://download.opensuse.org/repositories/isv:/Rancher:/Elemental:/Stable:/Teal53/media/iso/elemental-teal.x86_64.iso
         type: string
+      rancher_upgrade:
+        description: Rancher Manager channel/version to upgrade to
+        default: latest/devel
+        type: string
       rancher_version:
         description: Rancher Manager channel/version to use for installation
         default: stable/latest
@@ -39,10 +43,10 @@ jobs:
       destroy_runner: ${{ inputs.destroy_runner }}
       iso_to_test: ${{ inputs.iso_to_test }}
       k8s_version_to_provision: v1.25.7+rke2r1
-      # Force to only deploy the first 3 nodes
-      node_number: 3
+      node_number: 5
       operator_upgrade: oci://registry.opensuse.org/isv/rancher/elemental/${{ inputs.upgrade_os_channel }}/charts/rancher
       operator_repo: oci://registry.opensuse.org/isv/rancher/elemental/stable/charts/rancher
+      rancher_upgrade: ${{ inputs.rancher_upgrade }}
       rancher_version: ${{ inputs.rancher_version }}
       upgrade_channel_list: registry.opensuse.org/isv/rancher/elemental/dev/teal53/15.4/rancher/elemental-teal-dev-channel/5.3:latest
       upgrade_image: registry.opensuse.org/isv/rancher/elemental/${{ inputs.upgrade_os_channel }}/teal53/15.4/rancher/elemental-teal/5.3:latest

--- a/.github/workflows/master-e2e.yaml
+++ b/.github/workflows/master-e2e.yaml
@@ -89,6 +89,9 @@ on:
         description: Rancher Manager channel/version to use for installation
         default: stable/latest
         type: string
+      rancher_upgrade:
+        description: Rancher Manager channel/version to upgrade to
+        type: string
       runner_template:
         description: Runner template to use
         default: elemental-e2e-ci-runner-spot-x86-64-template-n2-standard-16-v4
@@ -265,9 +268,10 @@ jobs:
           CA_TYPE: ${{ inputs.ca_type }}
           OPERATOR_REPO: ${{ inputs.operator_repo }}
           PROXY: ${{ inputs.proxy }}
+          PUBLIC_DNS: ${{ needs.create-runner.outputs.public_dns }}
           PUBLIC_DOMAIN: bc.googleusercontent.com
           TEST_TYPE: ${{ inputs.test_type }}
-        run: cd tests && PUBLIC_DNS=${{ needs.create-runner.outputs.public_dns }} make e2e-install-rancher
+        run: cd tests && make e2e-install-rancher
       - name: Workaround for DynamicSchemas (if needed)
         run: |
           # Check if DynamicSchemas for MachineInventorySelectorTemplate exists
@@ -459,6 +463,24 @@ jobs:
           # Export values
           echo "operator_upgrade=${OPERATOR_UPGRADE}" >> ${GITHUB_OUTPUT}
           echo "operator_version=${OPERATOR_VERSION}" >> ${GITHUB_OUTPUT}
+      - name: Upgrade Rancher Manager
+        if: inputs.test_type == 'cli' && inputs.rancher_upgrade != ''
+        id: rancher_upgrade
+        env:
+          CA_TYPE: ${{ inputs.ca_type }}
+          PROXY: ${{ inputs.proxy }}
+          PUBLIC_DNS: ${{ needs.create-runner.outputs.public_dns }}
+          PUBLIC_DOMAIN: bc.googleusercontent.com
+          RANCHER_UPGRADE: ${{ inputs.rancher_upgrade }}
+        run: |
+          cd tests && make e2e-upgrade-rancher-manager
+          # Extract Rancher Manager version
+          RM_VERSION=$(kubectl get pod \
+                         --namespace cattle-system \
+                         -l app=rancher \
+                         -o jsonpath={.items[*].status.containerStatuses[*].image} 2> /dev/null || true)
+          # Export values
+          echo "rm_version=${RM_VERSION}" >> ${GITHUB_OUTPUT}
       - name: Upgrade node 1 to specified OS version with osImage
         if: inputs.test_type == 'cli' && inputs.upgrade_image != ''
         env:
@@ -577,6 +599,7 @@ jobs:
             echo "## Upgrade details" >> ${GITHUB_STEP_SUMMARY}
             echo "Elemental Operator Upgrade: ${{ steps.operator_upgrade.outputs.operator_upgrade }}" >> ${GITHUB_STEP_SUMMARY}
             echo "Elemental Operator Image: ${{ steps.operator_upgrade.outputs.operator_version }}" >> ${GITHUB_STEP_SUMMARY}
+            echo "Rancher Manager Image: ${{ steps.rancher_upgrade.outputs.rm_version }}" >> ${GITHUB_STEP_SUMMARY}
             echo "Channel list: ${{ inputs.upgrade_channel_list }}" >> ${GITHUB_STEP_SUMMARY}
             echo "Channel used: ${{ inputs.upgrade_os_channel }}" >> ${GITHUB_STEP_SUMMARY}
             echo "Upgrade image: ${{ inputs.upgrade_image }}" >> ${GITHUB_STEP_SUMMARY}

--- a/tests/Makefile
+++ b/tests/Makefile
@@ -82,8 +82,10 @@ e2e-ui-rancher: deps
 e2e-uninstall-operator:
 	ginkgo --label-filter uninstall-operator -r -v ./e2e
 e2e-upgrade-node: deps
-	ginkgo --label-filter upgrade -r -v ./e2e
+	ginkgo --label-filter upgrade-node -r -v ./e2e
 e2e-upgrade-operator: deps
 	ginkgo --label-filter upgrade-operator -r -v ./e2e
+e2e-upgrade-rancher-manager: deps
+	ginkgo --label-filter upgrade-rancher-manager -r -v ./e2e
 start-cypress-tests:
 	@./scripts/start-cypress-tests

--- a/tests/e2e/backup-restore_test.go
+++ b/tests/e2e/backup-restore_test.go
@@ -81,9 +81,8 @@ var _ = Describe("E2E - Install Backup/Restore Operator", Label("install-backup-
 		})
 
 		By("Waiting for rancher-backup-operator pod", func() {
-			// Wait for Pod to run
-			err := k.WaitForNamespaceWithPod("cattle-resources-system", "app.kubernetes.io/name=rancher-backup")
-			Expect(err).To(Not(HaveOccurred()))
+			// Wait for pod to be started
+			misc.CheckPod(k, [][]string{{"cattle-resources-system", "app.kubernetes.io/name=rancher-backup"}})
 		})
 	})
 })

--- a/tests/e2e/helpers/install/install.go
+++ b/tests/e2e/helpers/install/install.go
@@ -1,0 +1,86 @@
+/*
+Copyright © 2023 SUSE LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package install
+
+import (
+	"strings"
+
+	. "github.com/onsi/gomega"
+	"github.com/rancher-sandbox/ele-testhelpers/kubectl"
+)
+
+// Install or upgrade Rancher Manager
+func DeployRancherManager(hostname, channel, version, ca, proxy string) {
+	channelName := "rancher-" + channel
+
+	// Add Helm repository
+	err := kubectl.RunHelmBinaryWithCustomErr("repo", "add", channelName,
+		"https://releases.rancher.com/server-charts/"+channel,
+	)
+	Expect(err).To(Not(HaveOccurred()))
+
+	err = kubectl.RunHelmBinaryWithCustomErr("repo", "update")
+	Expect(err).To(Not(HaveOccurred()))
+
+	// Set flags for Rancher Manager installation
+	flags := []string{
+		"upgrade", "--install", "rancher", channelName + "/rancher",
+		"--namespace", "cattle-system",
+		"--create-namespace",
+		"--set", "hostname=" + hostname,
+		"--set", "extraEnv[0].name=CATTLE_SERVER_URL",
+		"--set", "extraEnv[0].value=https://" + hostname,
+		"--set", "extraEnv[1].name=CATTLE_BOOTSTRAP_PASSWORD",
+		"--set", "extraEnv[1].value=rancherpassword",
+		"--set", "replicas=1",
+		"--set", "global.cattle.psp.enabled=false",
+	}
+
+	// Set specified version if needed
+	if version != "" && version != "latest" {
+		if version == "devel" {
+			flags = append(flags,
+				"--devel",
+				"--set", "rancherImageTag=v2.7-head",
+			)
+		} else if strings.Contains(version, "-rc") {
+			flags = append(flags,
+				"--devel",
+				"--version", version,
+			)
+		} else {
+			flags = append(flags, "--version", version)
+		}
+	}
+
+	// For Private CA
+	if ca == "private" {
+		flags = append(flags,
+			"--set", "ingress.tls.source=secret",
+			"--set", "privateCA=true",
+		)
+	}
+
+	// Use Rancher Manager behind proxy
+	if proxy == "rancher" {
+		flags = append(flags,
+			"--set", "proxy=http://172.17.0.1:3128",
+			"--set", "noProxy=127.0.0.0/8\\,10.0.0.0/8\\,cattle-system.svc\\,172.16.0.0/12\\,192.168.0.0/16\\,.svc\\,.cluster.local",
+		)
+	}
+
+	err = kubectl.RunHelmBinaryWithCustomErr(flags...)
+	Expect(err).To(Not(HaveOccurred()))
+}

--- a/tests/e2e/helpers/misc/misc.go
+++ b/tests/e2e/helpers/misc/misc.go
@@ -472,3 +472,13 @@ func CreateTemp(baseName string) (string, error) {
 	}
 	return t.Name(), nil
 }
+
+func CheckPod(k *kubectl.Kubectl, checkList [][]string) error {
+	for _, check := range checkList {
+		if err := k.WaitForNamespaceWithPod(check[0], check[1]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}

--- a/tests/e2e/install_test.go
+++ b/tests/e2e/install_test.go
@@ -57,12 +57,14 @@ var _ = Describe("E2E - Install Rancher Manager", Label("install"), func() {
 					return err
 				}, misc.SetTimeout(2*time.Minute), 5*time.Second).Should(BeNil())
 			})
+
 			if clusterType == "hardened" {
 				By("Configuring hardened cluster", func() {
 					err := exec.Command("sudo", installHardenedScript).Run()
 					Expect(err).To(Not(HaveOccurred()))
 				})
 			}
+
 			By("Starting RKE2", func() {
 				err := exec.Command("sudo", "systemctl", "enable", "--now", "rke2-server.service").Run()
 				Expect(err).To(Not(HaveOccurred()))
@@ -82,14 +84,12 @@ var _ = Describe("E2E - Install Rancher Manager", Label("install"), func() {
 				err := os.Setenv("KUBECONFIG", "/etc/rancher/rke2/rke2.yaml")
 				Expect(err).To(Not(HaveOccurred()))
 
-				err = k.WaitForNamespaceWithPod("kube-system", "k8s-app=kube-dns")
-				Expect(err).To(Not(HaveOccurred()))
-
-				err = k.WaitForNamespaceWithPod("kube-system", "app=rke2-metrics-server")
-				Expect(err).To(Not(HaveOccurred()))
-
-				err = k.WaitForNamespaceWithPod("kube-system", "app.kubernetes.io/name=rke2-ingress-nginx")
-				Expect(err).To(Not(HaveOccurred()))
+				checkList := [][]string{
+					{"kube-system", "k8s-app=kube-dns"},
+					{"kube-system", "app=rke2-metrics-server"},
+					{"kube-system", "app.kubernetes.io/name=rke2-ingress-nginx"},
+				}
+				misc.CheckPod(k, checkList)
 
 				err = k.WaitLabelFilter("kube-system", "Ready", "rke2-ingress-nginx-controller", "app.kubernetes.io/name=rke2-ingress-nginx")
 				Expect(err).To(Not(HaveOccurred()))
@@ -111,12 +111,14 @@ var _ = Describe("E2E - Install Rancher Manager", Label("install"), func() {
 					return err
 				}, misc.SetTimeout(2*time.Minute), 5*time.Second).Should(BeNil())
 			})
+
 			if clusterType == "hardened" {
 				By("Configuring hardened cluster", func() {
 					err := exec.Command("sudo", installHardenedScript).Run()
 					Expect(err).To(Not(HaveOccurred()))
 				})
 			}
+
 			By("Starting K3s", func() {
 				err := exec.Command("sudo", "systemctl", "start", "k3s").Run()
 				Expect(err).To(Not(HaveOccurred()))
@@ -127,20 +129,14 @@ var _ = Describe("E2E - Install Rancher Manager", Label("install"), func() {
 
 			By("Waiting for K3s to be started", func() {
 				// Wait for all pods to be started
-				err := k.WaitForNamespaceWithPod("kube-system", "app=local-path-provisioner")
-				Expect(err).To(Not(HaveOccurred()))
-
-				err = k.WaitForNamespaceWithPod("kube-system", "k8s-app=kube-dns")
-				Expect(err).To(Not(HaveOccurred()))
-
-				err = k.WaitForNamespaceWithPod("kube-system", "k8s-app=metrics-server")
-				Expect(err).To(Not(HaveOccurred()))
-
-				err = k.WaitForNamespaceWithPod("kube-system", "app.kubernetes.io/name=traefik")
-				Expect(err).To(Not(HaveOccurred()))
-
-				err = k.WaitForNamespaceWithPod("kube-system", "svccontroller.k3s.cattle.io/svcname=traefik")
-				Expect(err).To(Not(HaveOccurred()))
+				checkList := [][]string{
+					{"kube-system", "app=local-path-provisioner"},
+					{"kube-system", "k8s-app=kube-dns"},
+					{"kube-system", "k8s-app=metrics-server"},
+					{"kube-system", "app.kubernetes.io/name=traefik"},
+					{"kube-system", "svccontroller.k3s.cattle.io/svcname=traefik"},
+				}
+				misc.CheckPod(k, checkList)
 			})
 		}
 
@@ -184,14 +180,12 @@ var _ = Describe("E2E - Install Rancher Manager", Label("install"), func() {
 				err = kubectl.RunHelmBinaryWithCustomErr(flags...)
 				Expect(err).To(Not(HaveOccurred()))
 
-				err = k.WaitForNamespaceWithPod("cert-manager", "app.kubernetes.io/component=controller")
-				Expect(err).To(Not(HaveOccurred()))
-
-				err = k.WaitForNamespaceWithPod("cert-manager", "app.kubernetes.io/component=webhook")
-				Expect(err).To(Not(HaveOccurred()))
-
-				err = k.WaitForNamespaceWithPod("cert-manager", "app.kubernetes.io/component=cainjector")
-				Expect(err).To(Not(HaveOccurred()))
+				checkList := [][]string{
+					{"cert-manager", "app.kubernetes.io/component=controller"},
+					{"cert-manager", "app.kubernetes.io/component=webhook"},
+					{"cert-manager", "app.kubernetes.io/component=cainjector"},
+				}
+				misc.CheckPod(k, checkList)
 			})
 		}
 
@@ -272,14 +266,13 @@ var _ = Describe("E2E - Install Rancher Manager", Label("install"), func() {
 				Expect(err).To(Not(HaveOccurred()))
 			}
 
-			err = k.WaitForNamespaceWithPod("cattle-system", "app=rancher")
-			Expect(err).To(Not(HaveOccurred()))
-
-			err = k.WaitForNamespaceWithPod("cattle-fleet-local-system", "app=fleet-agent")
-			Expect(err).To(Not(HaveOccurred()))
-
-			err = k.WaitForNamespaceWithPod("cattle-system", "app=rancher-webhook")
-			Expect(err).To(Not(HaveOccurred()))
+			// Wait for all pods to be started
+			checkList := [][]string{
+				{"cattle-system", "app=rancher"},
+				{"cattle-fleet-local-system", "app=fleet-agent"},
+				{"cattle-system", "app=rancher-webhook"},
+			}
+			misc.CheckPod(k, checkList)
 
 			// Check issuer for Private CA
 			if caType == "private" {
@@ -373,8 +366,8 @@ var _ = Describe("E2E - Install Rancher Manager", Label("install"), func() {
 				}, misc.SetTimeout(1*time.Minute), 10*time.Second).Should(BeNil())
 			}
 
-			err := k.WaitForNamespaceWithPod("cattle-elemental-system", "app=elemental-operator")
-			Expect(err).To(Not(HaveOccurred()))
+			// Wait for pod to be started
+			misc.CheckPod(k, [][]string{{"cattle-elemental-system", "app=elemental-operator"}})
 		})
 	})
 })

--- a/tests/e2e/install_test.go
+++ b/tests/e2e/install_test.go
@@ -24,6 +24,7 @@ import (
 	. "github.com/onsi/gomega"
 	"github.com/rancher-sandbox/ele-testhelpers/kubectl"
 	"github.com/rancher-sandbox/ele-testhelpers/tools"
+	"github.com/rancher/elemental/tests/e2e/helpers/install"
 	"github.com/rancher/elemental/tests/e2e/helpers/misc"
 )
 
@@ -190,63 +191,7 @@ var _ = Describe("E2E - Install Rancher Manager", Label("install"), func() {
 		}
 
 		By("Installing Rancher", func() {
-			err := kubectl.RunHelmBinaryWithCustomErr("repo", "add", "rancher",
-				"https://releases.rancher.com/server-charts/"+rancherChannel,
-			)
-			Expect(err).To(Not(HaveOccurred()))
-
-			err = kubectl.RunHelmBinaryWithCustomErr("repo", "update")
-			Expect(err).To(Not(HaveOccurred()))
-
-			// Set flags for Rancher Manager installation
-			flags := []string{
-				"upgrade", "--install", "rancher", "rancher/rancher",
-				"--namespace", "cattle-system",
-				"--create-namespace",
-				"--set", "hostname=" + rancherHostname,
-				"--set", "extraEnv[0].name=CATTLE_SERVER_URL",
-				"--set", "extraEnv[0].value=https://" + rancherHostname,
-				"--set", "extraEnv[1].name=CATTLE_BOOTSTRAP_PASSWORD",
-				"--set", "extraEnv[1].value=rancherpassword",
-				"--set", "replicas=1",
-				"--set", "global.cattle.psp.enabled=false",
-			}
-
-			// Set specified version if needed
-			if rancherVersion != "" && rancherVersion != "latest" {
-				if rancherVersion == "devel" {
-					flags = append(flags,
-						"--devel",
-						"--set", "rancherImageTag=v2.7-head",
-					)
-				} else if strings.Contains(rancherVersion, "-rc") {
-					flags = append(flags,
-						"--devel",
-						"--version", rancherVersion,
-					)
-				} else {
-					flags = append(flags, "--version", rancherVersion)
-				}
-			}
-
-			// For Private CA
-			if caType == "private" {
-				flags = append(flags,
-					"--set", "ingress.tls.source=secret",
-					"--set", "privateCA=true",
-				)
-			}
-
-			// Use Rancher Manager behind proxy
-			if proxy == "rancher" {
-				flags = append(flags,
-					"--set", "proxy=http://172.17.0.1:3128",
-					"--set", "noProxy=127.0.0.0/8\\,10.0.0.0/8\\,cattle-system.svc\\,172.16.0.0/12\\,192.168.0.0/16\\,.svc\\,.cluster.local",
-				)
-			}
-
-			err = kubectl.RunHelmBinaryWithCustomErr(flags...)
-			Expect(err).To(Not(HaveOccurred()))
+			install.DeployRancherManager(rancherHostname, rancherChannel, rancherVersion, caType, proxy)
 
 			// Inject secret for Private CA
 			if caType == "private" {

--- a/tests/e2e/suite_test.go
+++ b/tests/e2e/suite_test.go
@@ -52,38 +52,41 @@ const (
 )
 
 var (
-	arch                 string
-	backupRestoreVersion string
-	caType               string
-	CertManagerVersion   string
-	clusterName          string
-	clusterNS            string
-	clusterType          string
-	elementalSupport     string
-	emulateTPM           bool
-	rancherHostname      string
-	imageVersion         string
-	isoBoot              string
-	k8sUpstreamVersion   string
-	k8sVersion           string
-	numberOfVMs          int
-	operatorUpgrade      string
-	operatorRepo         string
-	osImage              string
-	poolType             string
-	proxy                string
-	rancherChannel       string
-	rancherLogCollector  string
-	rancherVersion       string
-	sequential           bool
-	testType             string
-	upgradeChannelList   string
-	upgradeImage         string
-	upgradeOsChannel     string
-	upgradeType          string
-	usedNodes            int
-	vmIndex              int
-	vmName               string
+	arch                  string
+	backupRestoreVersion  string
+	caType                string
+	CertManagerVersion    string
+	clusterName           string
+	clusterNS             string
+	clusterType           string
+	elementalSupport      string
+	emulateTPM            bool
+	rancherHostname       string
+	imageVersion          string
+	isoBoot               string
+	k8sUpstreamVersion    string
+	k8sVersion            string
+	numberOfVMs           int
+	operatorUpgrade       string
+	operatorRepo          string
+	osImage               string
+	poolType              string
+	proxy                 string
+	rancherChannel        string
+	rancherLogCollector   string
+	rancherVersion        string
+	rancherUpgrade        string
+	rancherUpgradeChannel string
+	rancherUpgradeVersion string
+	sequential            bool
+	testType              string
+	upgradeChannelList    string
+	upgradeImage          string
+	upgradeOsChannel      string
+	upgradeType           string
+	usedNodes             int
+	vmIndex               int
+	vmName                string
 )
 
 func CheckClusterState(ns, cluster string) {
@@ -156,6 +159,7 @@ var _ = BeforeSuite(func() {
 	proxy = os.Getenv("PROXY")
 	rancherLogCollector = os.Getenv("RANCHER_LOG_COLLECTOR")
 	rancherVersion = os.Getenv("RANCHER_VERSION")
+	rancherUpgrade = os.Getenv("RANCHER_UPGRADE")
 	seqString := os.Getenv("SEQUENTIAL")
 	testType = os.Getenv("TEST_TYPE")
 	upgradeChannelList = os.Getenv("UPGRADE_CHANNEL_LIST")
@@ -211,6 +215,13 @@ var _ = BeforeSuite(func() {
 		s := strings.Split(rancherVersion, "/")
 		rancherChannel = s[0]
 		rancherVersion = s[1]
+	}
+
+	// Extract Rancher Manager channel/version to upgrade
+	if rancherUpgrade != "" {
+		s := strings.Split(rancherUpgrade, "/")
+		rancherUpgradeChannel = s[0]
+		rancherUpgradeVersion = s[1]
 	}
 
 	// Start HTTP server

--- a/tests/e2e/uninstall-operator_test.go
+++ b/tests/e2e/uninstall-operator_test.go
@@ -118,8 +118,8 @@ var _ = Describe("E2E - Uninstall Elemental Operator", Label("uninstall-operator
 				Expect(err).To(Not(HaveOccurred()))
 			}
 
-			err := k.WaitForNamespaceWithPod("cattle-elemental-system", "app=elemental-operator")
-			Expect(err).To(Not(HaveOccurred()))
+			// Wait for pod to be started
+			misc.CheckPod(k, [][]string{{"cattle-elemental-system", "app=elemental-operator"}})
 		})
 
 		By("Creating a dumb MachineRegistration", func() {

--- a/tests/e2e/upgrade_test.go
+++ b/tests/e2e/upgrade_test.go
@@ -25,6 +25,7 @@ import (
 	. "github.com/onsi/gomega"
 	"github.com/rancher-sandbox/ele-testhelpers/kubectl"
 	"github.com/rancher-sandbox/ele-testhelpers/tools"
+	"github.com/rancher/elemental/tests/e2e/helpers/install"
 	"github.com/rancher/elemental/tests/e2e/helpers/misc"
 )
 
@@ -64,12 +65,57 @@ var _ = Describe("E2E - Upgrading Elemental Operator", Label("upgrade-operator")
 		// Delay few seconds before checking, needed because we may have 2 pods at the same time
 		time.Sleep(misc.SetTimeout(30 * time.Second))
 
-		err = k.WaitForNamespaceWithPod("cattle-elemental-system", "app=elemental-operator")
-		Expect(err).To(Not(HaveOccurred()))
+		// Wait for all pods to be started
+		misc.CheckPod(k, [][]string{{"cattle-elemental-system", "app=elemental-operator"}})
 	})
 })
 
-var _ = Describe("E2E - Upgrading node", Label("upgrade"), func() {
+var _ = Describe("E2E - Upgrading Rancher Manager", Label("upgrade-rancher-manager"), func() {
+	// Create kubectl context
+	// Default timeout is too small, so New() cannot be used
+	k := &kubectl.Kubectl{
+		Namespace:    "",
+		PollTimeout:  misc.SetTimeout(300 * time.Second),
+		PollInterval: 500 * time.Millisecond,
+	}
+
+	It("Upgrade Rancher Manager", func() {
+		// Get before-upgrade Rancher Manager version
+		getImageVersion := []string{
+			"get", "pod",
+			"--namespace", "cattle-system",
+			"-l", "app=rancher",
+			"-o", "jsonpath={.items[*].status.containerStatuses[*].image}",
+		}
+		versionBeforeUpgrade, err := kubectl.Run(getImageVersion...)
+		Expect(err).To(Not(HaveOccurred()))
+
+		// Upgrade Rancher Manager
+		install.DeployRancherManager(rancherHostname, rancherUpgradeChannel, rancherUpgradeVersion, caType, proxy)
+
+		// Wait for Rancher Manager to be running
+		checkList := [][]string{
+			{"cattle-system", "app=rancher"},
+			{"cattle-fleet-local-system", "app=fleet-agent"},
+			{"cattle-system", "app=rancher-webhook"},
+		}
+		misc.CheckPod(k, checkList)
+
+		// Check that all pods are using the same version
+		Eventually(func() int {
+			out, _ := kubectl.Run(getImageVersion...)
+			return len(strings.Fields(out))
+		}, misc.SetTimeout(3*time.Minute), 5*time.Second).Should(Equal(1))
+
+		// Get after-upgrade Rancher Manager version
+		// and check that it's different to the before-upgrade version
+		versionAfterUpgrade, err := kubectl.Run(getImageVersion...)
+		Expect(err).To(Not(HaveOccurred()))
+		Expect(versionAfterUpgrade).To(Not(Equal(versionBeforeUpgrade)))
+	})
+})
+
+var _ = Describe("E2E - Upgrading node", Label("upgrade-node"), func() {
 	var (
 		wg           sync.WaitGroup
 		value        string

